clang/lib/Sema/SemaInit.cpp

```diff
@@ -743,6 +743,12 @@ void InitListChecker::FillInEmptyInitForField(unsigned Init, FieldDecl *Field,
         ILE->updateInit(SemaRef.Context, Init, Filler);
       return;
     }
+
+    if (Field->hasAttr<ExplicitInitAttr>()) {
+      SemaRef.Diag(ILE->getExprLoc(), diag::warn_field_requires_explicit_init)
```

Comment:
Your other comment seems right.

`InitListChecker` should not produce any diagnostics when `VerifyOnly == true`, so we should only report the warning when the flag is false.